File: verifier/typescript/verify.ts
// verifier/typescript/verify.ts
//
// The TypeScript reference verifier for the HumanAccepted receipt format.
// Mirrors the Python reference verifier byte-for-byte on the canonical
// form, and uses TweetNaCl for the Ed25519 verify (the only third-party
// dependency). No HTTP, no async, no logging — pure function over
// (receipt, public_key) → verdict.
//
// Usage:
//   import { verify, VerificationKey } from "./verify.ts";
//   const result = verify(receipt, publicKey); // "valid" | "invalid: <reason>"
//
// Used by the conformance test runner (test_vectors.ts) to assert the
// TypeScript implementation matches the Python reference on every vector.

import { canonicalBytes } from "./canonical.ts";
import * as nacl from "tweetnacl";

export type VerificationKey = Uint8Array; // 32 bytes, raw Ed25519 public key

export type Verdict =
  | { valid: true; reason: "valid" }
  | { valid: false; reason: string };

/** Parse a public key from a hex string (optionally prefixed with "ed25519:")
 *  or a 32-byte Uint8Array. Throws on invalid input. */
export function publicKeyFromHexOrBytes(input: string | Uint8Array): VerificationKey {
  if (input instanceof Uint8Array) {
    if (input.length !== 32) {
      throw new Error(`Ed25519 public key must be 32 bytes, got ${input.length}`);
    }
    return input;
  }
  const s = input.trim();
  const hex = s.startsWith("ed25519:") ? s.slice("ed25519:".length) : s;
  if (!/^[0-9a-fA-F]+$/.test(hex)) {
    throw new Error(`public key not valid hex: ${hex}`);
  }
  const bytes = new Uint8Array(hex.length / 2);
  for (let i = 0; i < bytes.length; i++) {
    bytes[i] = parseInt(hex.slice(i * 2, i * 2 + 2), 16);
  }
  if (bytes.length !== 32) {
    throw new Error(`Ed25519 public key must be 32 bytes, got ${bytes.length}`);
  }
  return bytes;
}

const MAX_KEY_LENGTH = 255;
const KEY_PATTERN = /^[A-Za-z0-9._\-]+$/;

/** Verify a receipt. Returns a Verdict describing the outcome. */
export function verify(receipt: unknown, publicKey: VerificationKey): Verdict {
  if (typeof receipt !== "object" || receipt === null) {
    return { valid: false, reason: "receipt is not an object" };
  }
  const r = receipt as Record<string, unknown>;
  if (r.version !== 1) {
    return { valid: false, reason: `unsupported version: ${r.version}` };
  }
  const sigs = (r.signatures as Record<string, unknown> | undefined) ?? {};
  const sigStr = sigs.tenant_ed25519 as string | undefined;
  if (!sigStr) {
    return { valid: false, reason: "missing signatures.tenant_ed25519" };
  }
  if (typeof sigStr !== "string" || !sigStr.startsWith("ed25519:")) {
    return { valid: false, reason: `unexpected signature prefix: ${sigStr}` };
  }
  const sigHex = sigStr.slice("ed25519:".length);
  if (!/^[0-9a-fA-F]+$/.test(sigHex)) {
    return { valid: false, reason: "signature not valid hex" };
  }
  const sig = new Uint8Array(sigHex.length / 2);
  for (let i = 0; i < sig.length; i++) {
    sig[i] = parseInt(sigHex.slice(i * 2, i * 2 + 2), 16);
  }
  if (sig.length !== 64) {
    return { valid: false, reason: `Ed25519 signature must be 64 bytes, got ${sig.length}` };
  }

  // Sign over the canonical bytes of (receipt minus the signatures block).
  const body: Record<string, unknown> = {};
  for (const [k, v] of Object.entries(r)) {
    if (k !== "signatures") body[k] = v;
  }
  let message: Uint8Array;
  try {
    message = canonicalBytes(body);
  } catch (e) {
    return { valid: false, reason: `canonicalisation failed: ${(e as Error).message}` };
  }

  // Ed25519 verify via tweetnacl. Returns false on any mismatch.
  const ok = nacl.sign.detached.verify(message, sig, publicKey);
  if (!ok) {
    return { valid: false, reason: "tenant signature did not verify" };
  }

  return { valid: true, reason: "valid" };
}

// Suppress unused warnings: these are exported for callers who want to
// pre-parse a key.
export const __exported = { MAX_KEY_LENGTH, KEY_PATTERN };

File: verifier/python/verify.py
"""
verify.py — HumanAccepted receipt format v0.1.0 reference verifier.

Verifies a receipt against a tenant's Ed25519 public key, offline.
No network calls. No third-party dependencies besides the standard
library and the `cryptography` package.

Usage:
    python3 verify.py <receipt.json> <tenant_public_key_hex>

Exit code 0 if valid, 1 if invalid, 2 if input error.

The byte-exact canonical form, the Ed25519 signature scheme, and the
field requirements are all defined in docs/receipt-format.md. This file
is the byte-exact Python implementation of the rules in §3-§5 of the spec.
"""

import json
import sys
from typing import Any, Optional

from cryptography.exceptions import InvalidSignature
from cryptography.hazmat.primitives.asymmetric.ed25519 import Ed25519PublicKey
from cryptography.hazmat.primitives.serialization import load_der_public_key, load_pem_public_key

VERSION = "0.1.0"


def canonical_bytes(obj: Any) -> bytes:
    """Recursive-sorted-keys, keep null, drop undefined, compact JSON.
    Byte-exact match with the TypeScript reference and the Worker."""
    if obj is None:
        return b"null"
    if obj is True:
        return b"true"
    if obj is False:
        return b"false"
    if isinstance(obj, (int, float)):
        return json.dumps(obj, separators=(",", ":"), ensure_ascii=False).encode("utf-8")
    if isinstance(obj, str):
        return json.dumps(obj, separators=(",", ":"), ensure_ascii=False).encode("utf-8")
    if isinstance(obj, list):
        return b"[" + b",".join(canonical_bytes(x) for x in obj) + b"]"
    if isinstance(obj, dict):
        # Drop only undefined-like (None as a key value is a real null; keep it).
        # Filter out dict entries whose value is the literal "undefined" (shouldn't
        # happen in a real JSON load, but defensive).
        items = [(k, v) for k, v in obj.items() if v is not None or k in obj]
        items.sort(key=lambda kv: kv[0])
        body = b",".join(
            json.dumps(k, separators=(",", ":"), ensure_ascii=False).encode("utf-8")
            + b":"
            + canonical_bytes(v)
            for k, v in items
        )
        return b"{" + body + b"}"
    raise TypeError(f"unsupported type: {type(obj)}")


def _public_key_from_hex_or_pem(s: str) -> Ed25519PublicKey:
    s = s.strip()
    if "BEGIN" in s:
        if "PRIVATE" in s.upper():
            raise ValueError("private key supplied; verifier needs the public key only")
        return load_pem_public_key(s.encode("utf-8"))
    # Raw 32-byte hex, optionally prefixed "ed25519:"
    if s.startswith("ed25519:"):
        s = s[len("ed25519:") :]
    try:
        raw = bytes.fromhex(s)
    except ValueError as e:
        raise ValueError(f"public key not valid hex: {e}") from e
    if len(raw) != 32:
        raise ValueError(f"Ed25519 public key must be 32 bytes, got {len(raw)}")
    return Ed25519PublicKey.from_public_bytes(raw)


def verify(receipt: dict, public_key: Ed25519PublicKey) -> tuple[bool, str]:
    """Returns (ok, reason). reason is human-readable."""
    if not isinstance(receipt, dict):
        return False, "receipt is not an object"
    if receipt.get("version") != 1:
        return False, f"unsupported version: {receipt.get('version')!r}"
    sigs = receipt.get("signatures") or {}
    sig_hex = sigs.get("tenant_ed25519")
    if not sig_hex:
        return False, "missing signatures.tenant_ed25519"
    if not sig_hex.startswith("ed25519:"):
        return False, f"unexpected signature prefix: {sig_hex[:10]!r}"
    try:
        sig_bytes = bytes.fromhex(sig_hex[len("ed25519:") :])
    except ValueError as e:
        return False, f"signature not valid hex: {e}"
    if len(sig_bytes) != 64:
        return False, f"Ed25519 signature must be 64 bytes, got {len(sig_bytes)}"

    # Sign over the canonical bytes of (receipt minus the `signatures` block).
    body = {k: v for k, v in receipt.items() if k != "signatures"}
    try:
        msg = canonical_bytes(body)
    except Exception as e:  # pragma: no cover
        return False, f"canonicalisation failed: {e}"

    try:
        public_key.verify(sig_bytes, msg)
    except InvalidSignature:
        return False, "tenant signature did not verify"

    # Optional: also check the canonical form is byte-identical to a round-trip
    # (catches any field that wasn't in the original source).
    return True, "valid"


def main(argv: list[str]) -> int:
    if len(argv) != 3:
        print(f"usage: python3 verify.py <receipt.json> <public-key.pem-or-hex>", file=sys.stderr)
        return 2
    with open(argv[1], "r", encoding="utf-8") as f:
        receipt = json.load(f)
    try:
        pub = _public_key_from_hex_or_pem(argv[2])
    except ValueError as e:
        print(f"error: {e}", file=sys.stderr)
        return 2
    ok, reason = verify(receipt, pub)
    print(reason)
    return 0 if ok else 1


if __name__ == "__main__":
    sys.exit(main(sys.argv))
